### DISCUSSION TITLE
NextJS Head rendering client side

### DISCUSSION THREAD

**Discussion by Vanals** [vercel/next.js#33059] [Help]
### Run `next info` (available from version 12.0.8 and up)

N/A

### What version of Next.js are you using?

12.0.7

### What version of Node.js are you using?

v16.13.0

### What browser are you using?

Chrome

### What operating system are you using?

MacOS

### How are you deploying your application?

next

### Describe the Bug

Hi,

I am facing a similar issue reported in other tickets where the but could not find a solution and to be honest understand what the problem is. 

The issue is: In _app.js I am wrapping meta tags inside next/head but they gets rendered only client-side. The rest is rendered statically.

_app.js is a class react component with getInitialProps, and we are rendering statically with `next build && next export`. 

I also thought getInitialProps might not be the right one to use?  but In a separated clean react project I tried to replicate this setup (class + getInitialProps) and works fine the head is rendered statically too, but not in the project I am working on. So must be something else... ?

**A snippet of _app.tsx.**

```js
export default class MyApp extends App<Props, State> {

  ...LOGIC

  static async getInitialProps({Component, ctx}: AppContext) {
    let pageProps = {};
    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx);
    }
    return {path: ctx.asPath, pageProps};
  }

...LOGIC

render() {

    const {Component, pageProps, path} = this.props;
    const {theme, themeMode} = this.state as State;
    
    const handlers = [
      instrumentationHandlers.createConsoleHandler(),
      instrumentationHandlers.createTealiumHandler(),
    ];

    return (
      <>
        <Head>
          <title key='title'>TITLE</title>
          <meta
            key="description"
            name="description"
            content="DESCRIPTION"
          />
        </Head>
        <InstrumentationProvider
          {...createEventInstrumentation(handlers, {
            ...pageProps,
          })}
        >
          <ThemeProvider theme={theme}>
            <ThemeMode.Provider value={themeMode}>
              <Global
                styles={css`
                  html {
                    scroll-behavior: smooth;
                    scroll-padding-top: 90px;
                  }
                  @font-face {
                    font-family: 'DM Sans';
                    src: url('static/fonts/dmsans-regular-webfont.woff2')
                        format('woff2'),
                      url('static/fonts/dmsans-regular-webfont.woff')
                        format('woff');
                    font-style: normal;
                    font-weight: 400;
                    font-display: swap;
                  }
                  
              />

              <PageLoadInstrumentation />
              <Component
                {...pageProps}
                path={path}
                toggleTheme={this.toggleTheme}
                themeMode={themeMode}
              />
            </ThemeMode.Provider>
          </ThemeProvider>
        </InstrumentationProvider>
      </>
    );
  }
}
```

Rendering Head in _document.TSX works, it makes the Head statically rendered but I want it to be dynamic. So different for every page and. So ideally would like to have it in _app.tsx

**Other Things I tried:** 

We are also using Helmet in _document.js but commenting it out still, _app.js is rendered client-side.

Commenting everything apart from the meta tags in _app.js still is rendered CS

Thank you,

### Expected Behavior

<head tags should be rendered server-side and be caught by social networks.

### To Reproduce

I have been trying to reproduce it in a separated nextJS repo but couldn't.

**Comment by balazsorban44** [comment]
There is quite a lot going on in your `_app`, and I'm not sure how to reproduce it. It would help if you could isolate the issue and create a complete reproduction.

**Comment by Vanals** [comment]
As I said, I have been trying to reproduce it for debugging the issue, but I have not been able to reproduce it unfortunately. But Still, thought was worth raise the issue and hope someone could give an hand. I will update if I have any more info, but for now this is all.

**Comment by balazsorban44** [comment]
Using this very simple code:

```js
// pages/_app.jx
import Head from "next/head"

export default function MyApp({ Component, pageProps }) {
  return (
    <>
      <Head>
        <title key="title">TITLE</title>
        <meta key="description" name="description" content="DESCRIPTION" />
      </Head>
      <Component {...pageProps} />
    </>
  )
}
```

And running `next build && next export` produces an HTML containing the correct elements in `head`:

```js
// out/index.html
<!DOCTYPE html>
<html>

<head>
  <meta name="viewport" content="width=device-width" />
  <meta charSet="utf-8" />
  <title>TITLE</title>
  <meta name="description" content="DESCRIPTION" />
// ...
```

The problem seems to be specific for your use case so I converted it to a discussion.

**Comment by Vanals** [comment]
**Hi @balazsorban44  I managed to reproduce the issue.**
Could you please check the repo I created?
https://github.com/Vanals/Hello-NextJS

---
1) yarn

2) yarn next build && yarn next export

Outcome:
Meta tags from page2/index.tsx, _app.jsx and _document are in the static files (index.html & page2.html)

---

If you now go to `_document` and uncomment `getInitialProps()` and re run the step2 only the meta tags in _document will be rendered.

 ```
   static async getInitialProps(ctx) {
     const {html} = await ctx.renderPage();
     return {
       html,
     };
   }
 ```

We are currently using getInitialProps, and therefore renderPage(), in our repo to make a script work and we are also using a css-in-js library.

Is this a bug? some bad practice we are applying?

I would expect in both cases(commenting getInitialProps or not) to have all the meta tags, nested in `<Head>`, statically rendered.
As I understood from the docs getInitialProps does static rendering when exporting with `yarn next build && yarn next export`
Thank you

**Reply by balazsorban44** [reply]
Unfortunately, I still find the reproduction a bit hard to follow. Ideally, you would cut out any code from there that's unnecessary to reproduce the actual problem, and create a step-by-step way to reproduce the problem here in the description.

I also looked up our docs and:

https://nextjs.org/docs/advanced-features/static-html-export#getinitialprops

> getInitialProps will be called on every client-side navigation, if you'd like to only fetch data at build-time, switch to getStaticProps.

If you are doing `next export` (but in most cases really), the best solution might be using other data fetching methods instead: https://nextjs.org/docs/basic-features/data-fetching
